vowpalwabbit/core/include/vw/core/learner.h

```diff
@@ -1054,206 +665,361 @@ class common_learner_builder
   }
 };
 
-template <class DataT, class ExampleT, class BaseLearnerT>
+template <class DataT, class ExampleT>
 class reduction_learner_builder
-    : public common_learner_builder<reduction_learner_builder<DataT, ExampleT, BaseLearnerT>, DataT, ExampleT,
-          BaseLearnerT>
+    : public common_learner_builder<reduction_learner_builder<DataT, ExampleT>, DataT, ExampleT>
 {
 public:
-  using super =
-      common_learner_builder<reduction_learner_builder<DataT, ExampleT, BaseLearnerT>, DataT, ExampleT, BaseLearnerT>;
-  reduction_learner_builder(std::unique_ptr<DataT>&& data, BaseLearnerT* base, const std::string& name)
-      // NOTE: This is a copy of the base! The purpose is to copy all of the
+  using super = common_learner_builder<reduction_learner_builder<DataT, ExampleT>, DataT, ExampleT>;
+  reduction_learner_builder(std::unique_ptr<DataT>&& data, learner* prev, const std::string& name)
```

Comment:
`prev` being a shared_ptr here will allow us to propagate the lifetime all the way from he initial creation through the stack

Reply:
We use `std::enable_shared_from_this` to let us create a valid shared_ptr from a raw pointer. The code right now should already have correct lifetime of the shared_ptr from creation to finally getting stored in VW::workspace.

Reply:
It's clearer to the reader though if this were to express the shared ownership being passed. We should never use raw pointers for ownership which seems to be happening here even though it is being promoted to a shared pointer later

Reply:
Changed it to use shared_ptr now